### PR TITLE
Fix cargo_common_metadata warning on `publish = false`.

### LINKED ISSUE #6649: cargo_common_metadata shouldn't warn on `publish = false`
**Lint name:** cargo_common_metadata

**I tried this code:**
```rust
[package]
name = "cargo_common_metadata"
version = "0.1.0"
publish = false
```

**I expected to see this happen:**
I expect no errors to happen, because `publish` is set to false.
Sometimes you have crates in your workspace that you only use for testing or other purposes, which don't need any of the metadata, this is why they have `publish = false`.

**Instead, this happened:**
<details><summary>Warnings:</summary>

```
error: package `cargo_common_metadata` is missing `package.authors` metadata
   |
   = note: `-D clippy::cargo-common-metadata` implied by `-D warnings`

error: package `cargo_common_metadata` is missing `package.description` metadata

error: package `cargo_common_metadata` is missing `either package.license or package.license_file` metadata

error: package `cargo_common_metadata` is missing `package.repository` metadata

error: package `cargo_common_metadata` is missing `package.readme` metadata

error: package `cargo_common_metadata` is missing `package.keywords` metadata

error: package `cargo_common_metadata` is missing `package.categories` metadata

error: aborting due to 7 previous errors
```
</details>

### Meta

- `cargo clippy -V`: e.g. clippy 0.0.212 (f455e46 2020-06-20)
- `rustc -Vv`:
  ```
  rustc 1.51.0-nightly (e38fb306b 2021-01-14)
  binary: rustc
  commit-hash: e38fb306b7f5e65cca34df2dab1f0db15e1defb4
  commit-date: 2021-01-14
  host: x86_64-unknown-linux-gnu
  release: 1.51.0-nightly
  LLVM version: 11.0
  ```

### PR DESCRIPTION
**Pull request by daxpedda**
I believe `cargo_common_metadata` shouldn't trigger when `publish = false`, not sure if everybody agrees.
Made some tests to handle all edge-cases.

Fixes #6649.

changelog: [`cargo_common_metadata`](https://rust-lang.github.io/rust-clippy/master/index.html#cargo_common_metadata): No longer lints if [`publish = false`](https://doc.rust-lang.org/cargo/reference/manifest.html#the-publish-field) is defined in the manifest

### CODE REVIEW

File: tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,5 @@
 [package]
 name = "cargo_common_metadata"
 version = "0.1.0"
-publish = false
```
Comment (daxpedda) [inline tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml:4]: Is removing this gonna be an issue when publishing Clippy?

Comment (flip1995) [reply tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml:4]: So these are there because of https://github.com/rust-lang/rust-clippy/pull/5614#discussion_r427308121

> Might be worth adding `publish = false` to avoid accidental publishing

So I'd rather keep it. You can keep the current tests working by disabling this check with checking for  the env var `__CLIPPY_INTERNAL_TESTS=true`. But then the test for `publish = false` won't work...

Maybe add a configuration option, where this can be configured to the lint. See https://github.com/rust-lang/rust-clippy/blob/master/doc/adding_lints.md#adding-configuration-to-a-lint on how to do this.

Comment (daxpedda) [reply tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml:4]: I would rather not add a lint configuration option considering this is only useful for testing Clippy itself and the Clippy configuration is a user-facing interface.

But I have a different suggestion, not adding a "description" and "license" makes it impossible to publish to Crates.io, so I can document this in a comment inside `Cargo.toml` to make sure nobody adds it on accident.

What do you think?

Comment (flip1995) [reply tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml:4]: The problem then still is, that for the `pass` tests it will always pass, just because they have the `publish = false`, not because they actually pass the tests.

You can hide the configuration from the user, by prefacing the lint name with `_`. Then it won't appear in the documentation.

Comment (daxpedda) [reply tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml:4]: > The problem then still is, that for the `pass` tests it will always pass, just because they have the `publish = false`, not because they actually pass the tests.

I am afraid I don't understand what you are trying to tell me here, the `pass` tests don't always pass because they have `publish = false`, I just tried adding a `main.stderr` file in them and the test failed correctly.
Am I misunderstanding something here?

Comment (daxpedda) [reply tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml:4]: So I believe I know what you meant now: the `pass` test fullfills all criterias to upload to Crates.io but doesn't have `publish = false`.

So I did implement your suggestion with the configuration option, but I can't make the test pick up the `clippy.toml` though, I'm assuming `ui-cargo` tests don't pick up on it.

Any help would be appreciated.

File: clippy_lints/src/utils/conf.rs [hunk 2]
```diff
@@ -173,6 +173,8 @@ define_Conf! {
     (disallowed_methods, "disallowed_methods": Vec<String>, Vec::<String>::new()),
     /// Lint: UNREADABLE_LITERAL. Should the fraction of a decimal be linted to include separators.
     (unreadable_literal_lint_fractions, "unreadable_literal_lint_fractions": bool, true),
+    /// Lint: CARGO_COMMON_METADATA. For internal testing only, ignores the current `publish` settings in the Cargo manifest.
```
Comment (flip1995) [inline clippy_lints/src/utils/conf.rs:176]:
```suggestion
    /// Lint: _CARGO_COMMON_METADATA. For internal testing only, ignores the current `publish` settings in the Cargo manifest.
```

You will have to prefix the lint name to hid it in the documentation.

File: clippy_lints/src/cargo_common_metadata.rs [hunk 3]
```diff
@@ -80,32 +93,34 @@ impl LateLintPass<'_> for CargoCommonMetadata {
...
-                missing_warning(cx, &package, "package.readme");
-            }
-
-            if is_empty_vec(&package.keywords) {
-                missing_warning(cx, &package, "package.keywords");
-            }
-
-            if is_empty_vec(&package.categories) {
-                missing_warning(cx, &package, "package.categories");
+            if package.publish.as_ref().filter(|publish| publish.is_empty()).is_none() || self.ignore_publish {
```
Comment (flip1995) [inline clippy_lints/src/cargo_common_metadata.rs:96]: Shouldn't you check if `publish = false`? A check for `publish.is_none()` would bean, that it should only lint, if no publish attribute exists. So `publish = true` would also not get linted.

File: tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml (hunk 1 above)
Comment (flip1995) [reply tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml:4]: So your check prevents this lint from triggering if `publish = false`. All `ui-cargo` tests have `publish = false` in their `Cargo.toml`. For tests which should trigger a lint (`fail` tests, with a `stderr` file), won't trigger this lint anymore with your check for `publish = false` and you see this in their tests failing because of a different `stderr`. But the same problem exists for tests that **shouldn't** trigger a lint (`pass` tests, no `stderr` file). Before this change they didn't produce a `stderr`, **because no lint triggered on the test**. But with your change the test behavior silently changed: The `pass` tests don't emit a `stderr` **because they have `publish = false` in them**.

So what you need to do is to add a `clippy.toml` file to every `ui-cargo` test with `_cargo_ignore_publish = true` in it, except the tests where you're testing your change.

---



> So I did implement your suggestion with the configuration option, but I can't make the test pick up the `clippy.toml` though, I'm assuming `ui-cargo` tests don't pick up on it.

The tests should pick up the clippy.toml file. But you have to add it to all tests, **except** the tests you add for your change.

---

I hope I explained this well enough now. This is understandably confusing.

**Review by flip1995 — CHANGES_REQUESTED** [review]

File: clippy_lints/src/cargo_common_metadata.rs (hunk 3 above)
Comment (daxpedda) [reply clippy_lints/src/cargo_common_metadata.rs:96]: This check isn't really easy to read. I will add some documentation.
[The documentation says](https://docs.rs/cargo_metadata/0.12.3/cargo_metadata/struct.Package.html#structfield.publish):
> Publishing is unrestricted if None, and forbidden if the Vec is empty.

- Cases we want to lint for, because they allow publishing:
  - nothing -> `None`
  - `publish = true` -> `None`
  - `publish = ["somewhere"]` -> `Some(vec!["somewhere"])`
- Cases we do **NOT** want to lint for, because they do **NOT** allow publishing:
  - `publish = false` -> `Some(vec![])`
  - `publish = []` -> `Some(vec![])`

So basically we want to lint if it's none or the `Vec` isn't empty.

File: tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml (hunk 1 above)
Comment (daxpedda) [reply tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml:4]: So I only used the workaround with `cargo-ignore-publish` on tests that actually need `publish = false`, but some tests don't need it because they don't fulfill the requirements to be published anyway, like a missing `author` or `description`.

In any case, I added `publish = false` back in, but how do I test `publish = true`? As far as I can see it's not a problem, because as mentioned above, it's not possible to publish this crate anyway.
> The tests should pick up the clippy.toml file.

As far as I tested, it doesn't pick it up, I solved it by adding `CLIPPY_CONF_DIR` to the testsuite, I'm open to any other suggestion.

File: clippy_lints/src/cargo_common_metadata.rs (hunk 3 above)
Comment (daxpedda) [reply clippy_lints/src/cargo_common_metadata.rs:98]: https://github.com/rust-lang/rust-clippy/blob/fd8b5fa1aa614dcfd21471ab5a9a56c7873ce14a/clippy_lints/src/cargo_common_metadata.rs#L96-L98

I hope this is alright?

File: tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml (hunk 1 above)
Comment (daxpedda) [reply tests/ui-cargo/cargo_common_metadata/fail/Cargo.toml]: Everything seems to work just fine now.

File: tests/ui-toml/toml_unknown_key/conf_unknown_key.stderr [hunk 4]
```diff
@@ -1,4 +1,4 @@
...
+error: error reading Clippy's configuration file `$DIR/clippy.toml`: unknown field `foobar`, expected one of `msrv`, `blacklisted-names`, `cognitive-complexity-threshold`, `cyclomatic-complexity-threshold`, `doc-valid-idents`, `too-many-arguments-threshold`, `type-complexity-threshold`, `single-char-binding-names-threshold`, `too-large-for-stack`, `enum-variant-name-threshold`, `enum-variant-size-threshold`, `verbose-bit-mask-threshold`, `literal-representation-threshold`, `trivial-copy-size-limit`, `pass-by-value-size-limit`, `too-many-lines-threshold`, `array-size-threshold`, `vec-box-size-threshold`, `max-trait-bounds`, `max-struct-bools`, `max-fn-params-bools`, `warn-on-all-wildcard-imports`, `disallowed-methods`, `unreadable-literal-lint-fractions`, `cargo-ignore-publish`, `third-party` at line 5 column 1
```
Comment (flip1995) [inline tests/ui-toml/toml_unknown_key/conf_unknown_key.stderr:1]: This is not optimal, because it makes this option visible to the user. Nothing we can do here though, jjust for future reference. (But the error message is also horrible, so it's not that bad)

**Review by flip1995 — APPROVED** [review]

**Comment by flip1995** [comment]
@bors r+

Thanks! (Also for explaining the `publish` field to me.TIL)
